tools/CopyrightDateTool.scala

```diff
@@ -0,0 +1,174 @@
+import sys.process._
+
+object CopyrightDateTool {
+  val usage = """
+
+    This tool compares a copyright date in files with the date of last change in git.
+    If the message of last change contains "[Sync copyright dates]" it will be ignored and 
+    we will take next commit.
+
+    analyze - just shows if anything is ok
+    update - file fixing
+
+    From root folder:
+    Usage: scala tools/CheckDates.scala analyze
+           scala tools/CheckDates.scala update
+
+  """
+  trait DateComparison {
+    def fileLastUpdateYear: String
+    def gitLastUpdateYear: String
+    def fileCreationYear: String
+    def gitCreationYear: String
+
+    def isUpdateDateWrong(): Boolean = {
+      fileLastUpdateYear.toInt != gitLastUpdateYear.toInt
+    }
+
+    def isUpdateDateWrongStr(): Boolean = {
+      fileLastUpdateYear != gitLastUpdateYear
+    }
+  }
+
+  case class FileDates(pathToFile: String,
+                      fileLastUpdateYear: String,
+                      gitLastUpdateYear: String,
+                      fileCreationYear: String,
+                      gitCreationYear: String) extends DateComparison
+
+  case class GitInfo(message: String, date: String)
+
+  def readFilesList(): Array[String] = {
+    val git_files = (s"git ls-files") !!
```

Comment:
In fact we can go further and delegate the content matching to `git` as well,

From `git help grep`,

      git grep 'time_t' -- '*.[ch]'
           Looks for time_t in all tracked .c and .h files in the working directory and its subdirectories.

This will result in a list of files that are known to the repo, include the target string and end in .scala.

    git grep --name-only --fixed-strings "|    (c) 20" -- '*.scala'

For me this results in 602 files. Here is the output followed by a cross checking using `fgrep` and `find`,

    $ git grep --name-only --fixed-strings "|    (c) 20" -- '*.scala'|sort|cat -n|tail
    593	src/reflect/scala/reflect/io/File.scala
    594	src/scaladoc/scala/tools/ant/Scaladoc.scala
    595	test/disabled/pos/spec-List.scala
    596	test/files/pos/spec-Function1.scala
    597	test/instrumented/library/scala/runtime/ScalaRunTime.scala
    598	test/junit/scala/util/control/ExceptionTest.scala
    599	test/scaladoc/resources/doc-root/AnyRef.scala
    600	test/scaladoc/resources/doc-root/Any.scala
    601	test/scaladoc/resources/doc-root/Nothing.scala
    602	test/scaladoc/resources/doc-root/Null.scala

    $ fgrep -l  '|    (c) 20' $(find -name '*.scala')|sort|cat -n|tail
    593	./src/reflect/scala/reflect/io/File.scala
    594	./src/scaladoc/scala/tools/ant/Scaladoc.scala
    595	./test/disabled/pos/spec-List.scala
    596	./test/files/pos/spec-Function1.scala
    597	./test/instrumented/library/scala/runtime/ScalaRunTime.scala
    598	./test/junit/scala/util/control/ExceptionTest.scala
    599	./test/scaladoc/resources/doc-root/AnyRef.scala
    600	./test/scaladoc/resources/doc-root/Any.scala
    601	./test/scaladoc/resources/doc-root/Nothing.scala
    602	./test/scaladoc/resources/doc-root/Null.scala

By starting with a list of files known to pass this test,

    if(readFile(file_path).contains("|    (c) 20")) {

the number of files processed will be reduced by 10x so the execution time should drop down to ~ 30 seconds.